### PR TITLE
Proposal for improved modeling of creatives and ad formats

### PR DESCRIPTION
**Pull request by bokelley**
Introduces concepts of placement and ad format; adds device coverage-seconds metric; standardizes payload calculation for media files

### CODE REVIEW

File: docs/creative.mdx [hunk 1]
```diff
@@ -1,87 +1,81 @@
-# Creative
+# Measuring the carbon footprint of creatives and ad formats
+
+## Terminology: Placements, Ad Formats, Creatives, and Assets
+
+A _placement_ is an empty spot on a particular media property where an ad could be inserted. An "ad hole" if you will! Examples - a rectangle on a web page; the interstitial between levels of a game, the airtime before a podcast starts, a wall at an airport. In an _embedded_ placement (web page, social news feed) the ad sits inside content that the user controls. In an _ad-primary_ placement the creative IS the content and controls duration.
+
+A _media asset_ is an image, video, audio, or text that the advertiser wants to deliver to the consumer. A _creative template_ or _ad format_ is code, metadata, or other _non-media assets_ that, in combination with one or more media assets, faciliate the delivery of the advertisement to the consumer.
```
Comment (EmmaLouise2018) [inline docs/creative.mdx:7]: Do we need both creative template and ad format here?

File: docs/creative.mdx [hunk 2]
```diff
@@ -1,87 +1,81 @@
...
+
+A _creative_ is one or more media assets combined with an ad format. For instance, a CTV creative would typically be a 15-second or 30-second video (the media asset) wrapped in VAST/VPAID (the ad format) and deployed in a commercial break (the placement).
+
+## Defaults
```
Comment (EmmaLouise2018) [inline docs/creative.mdx:11]: We should add default placement ad primary and average view s per channel here

File: docs/calculations.mdx [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+| `creative_audio_duration_seconds`    | Audio duration in seconds |
+| `creative_video_data_transfer_bytes` | Data transfer for the video itself , ideally measured by the CDN |
+| `creative_video_vast_bytes`          | Size of VAST/VPAID wrapper for the video in bytes |
+| `creative_video_bitrate_kbps`        | Video bitrate *as delivered* in kilobits per second |
+| `creative_video_size_bytes`          | Video size *as delivered* in bytes |
+| `creative_video_duration_seconds`    | Video duration in seconds |
+| `creative_video_view_time`           | Average time, in seconds, that a video is viewed |
+| `creative_video_view_rate`           | Average percentage of a video that is viewed |
+
+## Creative Data Transfer
```
Comment (EmmaLouise2018) [inline docs/calculations.mdx:70]: Do we want to outline `Creative Consumer Device` as well?

File: docs/calculations.mdx [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+Bitrate (choose first to match)
+
+`bitrate = creative_video_bitrate_kbps`
+
+`bitrate = creative_video_size_bytes / 8 / creative_video_duration / 1024`
+
+`bitrate = default_video_bitrate[device_type]`
+
+Seconds streamed (choose first to match)
```
Comment (EmmaLouise2018) [inline docs/calculations.mdx:120]: Is seconds streamed also used to the power model calculation? Should we call out the power vs conventional modal pieces?

File: docs/creative.mdx [hunk 5]
```diff
@@ -1,87 +1,90 @@
-# Creative
+# Measuring the carbon footprint of creatives and ad formats
+
+## Terminology: Placements, Ad Formats, Creatives, and Assets
+
+A _placement_ is an empty spot on a particular media property where an ad could be inserted. An "ad hole" if you will! Examples - a rectangle on a web page; the interstitial between levels of a game, the airtime before a podcast starts, a wall at an airport. In an _embedded_ placement (web page, social news feed) the ad sits inside content that the user controls. In an _ad-primary_ placement the creative IS the content and controls duration.
+
+A _media asset_ is an image, video, audio, or text that the advertiser wants to deliver to the consumer. An _ad format_ is code, metadata, or other _non-media assets_ that, in combination with one or more media assets, faciliate the delivery of the advertisement to the consumer.
```
Comment (EmmaLouise2018) [inline docs/creative.mdx:7]: typo: `faciliate` => facilitate

File: docs/calculations.mdx [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+| `country`                            | The country where the impression was delivered |
+| `region`                             | The region of the country where the impression was delivered |
+| `channel`                            | See [channels](#channels) |
+| `device_type`                        | See [device types](#device-types) |
+| `network_type`                       | See [network types](#network-types) |
+| `creative_ad_format`                 | Either a basic or vendor-provided ad format |
+| `creative_image_data_transfer_bytes` | Data transfer for the video itself , ideally measured by the CDN |
+| `creative_image_width_pixels`        | Width of creative (basic image only) |
+| `creative_image_height_pixels`       | Height of creative (basic image only) |
+| `creative_audio_data_transfer_bytes` | Data transfer for the video itself , ideally measured by the CDN |
```
Comment (EmmaLouise2018) [inline docs/calculations.mdx:116]: Typo: Data transfer for audio - not video.

File: docs/snippets/defaults_consumer_device.mdx [hunk 7]
```diff
@@ -0,0 +1,26 @@
+```
+default_device_watts:
+  scope3:
```
Comment (EmmaLouise2018) [inline docs/snippets/defaults_consumer_device.mdx:3]: Probably should then not have then also duplicated in `templates/end_user_devices`

Comment (bokelley) [reply docs/snippets/defaults_consumer_device.mdx:3]: agree - was going to zap all of the overlapping defaults but wasn't sure how we wanted to handle the migration. Thought maybe we'd flip the API over to the new ones first?

**Review by EmmaLouise2018 — APPROVED** [review]

File: docs/calculations.mdx [hunk 8]
```diff
@@ -0,0 +1,422 @@
...
+      creative_video_bitrate_kbps ??
+      creative_video_size_bytes / 8 / creative_video_duration / 1024 ??
+      default_bitrate
+video_duration = creative_video_duration_seconds ?? ad_format.video_duration_seconds
+seconds_watched = creative_video_view_time ??
+                  creative_video_view_rate x video_duration ??
+                  video_duration
+buffer = ad_format.video_player.buffering_seconds ?? infinity
+seconds_streamed = MAX(seconds_watched + buffer, video_duration)
+loads = ad_format.video_player.lazy_loads ? views : impressions
```
Comment (bokelley) [inline docs/calculations.mdx:274]: @pymble2073 should we add another conditional here for DOOH loads? and should the delivery row have loads? I guess we need this for images too

File: docs/snippets/defaults_consumer_device.mdx (hunk 7 above)
Comment (EmmaLouise2018) [reply docs/snippets/defaults_consumer_device.mdx:3]: Yeah. There are things relying on it for sure. Let's merge this and then we'll convert the pipeline and then nuke it.

File: docs/calculations.mdx [hunk 9]
```diff
@@ -0,0 +1,422 @@
...
+Otherwise, calculate data transfer:
+
+```
+image_bytes = SUM(image_sizes[n].width x image_sizes[n].height) x 3 / default_image_compression_ratio
+
+audio_duration = creative_audio_duration_seconds ?? ad_format.audio_duration_seconds
+audio_bytes = audio_duration_seconds x default_audio_bitrate[device_type]
+
+default_bitrate =
+      ad_format.is_primary_experience ?
```
Comment (EmmaLouise2018) [inline docs/calculations.mdx:261]: @bokelley If we don't have an ad format - should we default this based on the channel?

Comment (bokelley) [reply docs/calculations.mdx:261]: I think in this context we would always have an ad format. From a spec perspective should we define the default ad format for each channel or is that an implementation choice?

**Pull request merged**